stdlib/http/src/main/java/org/ballerinalang/net/http/compiler/ResourceValidator.java

```diff
@@ -2,68 +2,172 @@
+                    break;
+                default:
+                    // do not execute
+            }
+        }
+    }
+
+    private static boolean isInvalidResourceParam(BLangSimpleVariable param, String expectedType) {
+        return !expectedType.equals(param.type.toString());
+    }
+
+    private static String getCompatibleAnnotation(BLangSimpleVariable param) {
+        for (BLangAnnotationAttachment annotationAttachment : param.annAttachments) {
+            BAnnotationSymbol annotationSymbol = annotationAttachment.annotationSymbol;
+            if (!PROTOCOL_HTTP.equals(annotationSymbol.pkgID.name.value)) {
+                continue;
+            }
+            String annotationName = annotationSymbol.name.value;
+            if (ANN_NAME_PATH_PARAM.equals(annotationName) || ANN_NAME_QUERY_PARAM.equals(annotationName) ||
+                    ANN_NAME_BODY_PARAM.equals(annotationName)) {
+                return annotationName;
+            }
+        }
+        return null;
+    }
+
+    private static void validatePathParam(BLangSimpleVariable param, List<String> pathSegments, DiagnosticLog dlog) {
+        BType paramType = param.type;
+        int varTag = paramType.tag;
+
+        if (varTag != TypeTags.STRING && varTag != TypeTags.INT && varTag != TypeTags.BOOLEAN &&
+                varTag != TypeTags.FLOAT) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible path param type: `" + param.type +
+                    "`, expected `string`, `int`, `boolean`, `float`");
+            return;
+        }
+        if (!pathSegments.contains(param.name.value)) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "invalid path param: `" + param.toString() +
+                    "`, missing segment `{" + param.name.value + "}` in the path config of the resource annotation");
+        }
+    }
+
+    private static void validateQueryParam(BLangSimpleVariable param, DiagnosticLog dlog) {
+        BType paramType = param.type;
+        int varTag = paramType.tag;
+
+        if (varTag != TypeTags.STRING && (varTag != TypeTags.ARRAY ||
+                ((BArrayType) paramType).getElementType().tag != TypeTags.STRING)) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible query param type: `" + param.type +
+                    "`, expected `string`, `string[]`");
+        }
+    }
+
+    private static void validateBodyParam(BLangSimpleVariable param, DiagnosticLog dlog) {
+        BType paramType = param.type;
+        int type = paramType.tag;
+
+        if (type != TypeTags.RECORD && type != TypeTags.JSON && type != TypeTags.XML &&
+                type != TypeTags.STRING && (type != TypeTags.ARRAY || !validArrayType(paramType))) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible entity-body param type: `"
+                    + param.type + "`, expected `string`, `json`, `xml`, `byte[]`, `{}`, `{}[]`");
```

Comment:
Same comment as above here, regarding error message format, use of unions and formatting